flycontrols.js

```diff
@@ -0,0 +1,228 @@
+
+      quat.mult(
+        flycontrols.currentYaw,
+        quat.fromAxisAngle(
+          quat.create(),
+          [0, 1, 0],
+          -movementX * flycontrols.sensitivityX
+        )
+      )
+      quat.mult(
+        flycontrols.currentPitch,
+        quat.fromAxisAngle(
+          quat.create(),
+          [1, 0, 0],
+          -movementY * flycontrols.sensitivityY
+        )
+      )
+
+      let newRot = quat.mult(
+        quat.mult([0, 0, 0, 1], flycontrols.currentYaw),
+        flycontrols.currentPitch
+      )
+
+      flycontrols.entity.transform.set({ rotation: newRot })
+    }
+  }
+
+  function onMouseUp(e) {
+    if (!flycontrols.enabled) return
+    if (flycontrols.releaseOnMouseUp) {
+      document.exitPointerLock()
+      flycontrols.moveForward = false
+      flycontrols.moveLeft = false
+      flycontrols.moveBackward = false
+      flycontrols.moveRight = false
+      flycontrols.moveUp = false
+      flycontrols.moveDown = false
+    }
+  }
+
+  function onKeyDown(e) {
+    if (!flycontrols.enabled) return
+    if (!document.pointerLockElement) return
+
+    e.preventDefault()
+
+    switch (e.keyCode) {
+      case 38: /*up*/
+      case 87:
+        /*W*/ flycontrols.moveForward = true
+        break
+
+      case 37: /*left*/
+      case 65:
+        /*A*/ flycontrols.moveLeft = true
+        break
+
+      case 40: /*down*/
+      case 83:
+        /*S*/ flycontrols.moveBackward = true
+        break
+
+      case 39: /*right*/
+      case 68:
+        /*D*/ flycontrols.moveRight = true
+        break
+
+      case 32:
+        /*space*/ flycontrols.moveUp = true
+        break
+
+      case 16:
+        /*shift*/ flycontrols.moveDown = true
+        break
+    }
+  }
+
+  function onKeyUp(e) {
+    if (!flycontrols.enabled) return
+    if (!document.pointerLockElement) return
+
+    switch (e.keyCode) {
+      case 38: /*up*/
+      case 87:
+        /*W*/ flycontrols.moveForward = false
+        break
+
+      case 37: /*left*/
+      case 65:
+        /*A*/ flycontrols.moveLeft = false
+        break
+
+      case 40: /*down*/
+      case 83:
+        /*S*/ flycontrols.moveBackward = false
+        break
+
+      case 39: /*right*/
+      case 68:
+        /*D*/ flycontrols.moveRight = false
+        break
+
+      case 32:
+        /*space*/ flycontrols.moveUp = false
+        break
+
+      case 16:
+        /*shift*/ flycontrols.moveDown = false
+        break
+    }
+  }
+
+  this._onMouseDown = onMouseDown
+  this._onMouseMove = onMouseMove
+  this._onMouseUp = onMouseUp
+
+  this._onKeyUp = onKeyUp
+  this._onKeyDown = onKeyDown
+
+  this.element.addEventListener('mousedown', onMouseDown)
+
+  document.addEventListener('mousemove', onMouseMove)
+  document.addEventListener('mouseup', onMouseUp)
+
+  document.addEventListener('keydown', onKeyDown)
+  document.addEventListener('keyup', onKeyUp)
+}
+
+FlyControls.prototype.dispose = function() {
```

Comment:
remove 'touchend' from where? Which raise the question: is this gonna work on touch devices?